cmd_migrate_db.go

```diff
@@ -0,0 +1,705 @@
+		destDb.Close()
+
+	}
+
+	logger.Info("!!!Migration of all mandatory db parts completed " +
+		"successfully!!!")
+
+	return nil
+}
+
+// validateDBBackends ensures that only migrations from bolt to sqlite/postgres
+// are allowed.
+func (x *migrateDBCommand) validateDBBackends() error {
+	// Source must be bolt
+	if x.Source.Backend != lncfg.BoltBackend {
+		return fmt.Errorf("source database must be bolt, got: %s",
+			x.Source.Backend)
+	}
+
+	// Destination must be sqlite or postgres.
+	switch x.Dest.Backend {
+	case lncfg.SqliteBackend, lncfg.PostgresBackend:
+		return nil
+	default:
+		return fmt.Errorf("destination database must be sqlite or "+
+			"postgres, got: %s", x.Dest.Backend)
+	}
+}
+
+// openSourceDb opens the source database.
+func openSourceDb(cfg *SourceDB, prefix, network string) (walletdb.DB, error) {
+	path := getBoltDBPath(cfg, prefix, network)
+	if path == "" {
+		return nil, fmt.Errorf("unknown prefix: %s", prefix)
+	}
+
+	const (
+		noFreelistSync = true
+		timeout        = time.Minute
+	)
+
+	args := []interface{}{
+		path, noFreelistSync, timeout,
+	}
+	backend := kvdb.BoltBackendName
+	logger.Infof("Opening bolt backend at %s for prefix '%s'",
+		path, prefix)
+
+	return kvdb.Open(backend, args...)
+}
+
+// openDestDb opens the different types of databases.
+func openDestDb(cfg *DestDB, prefix, network string) (walletdb.DB, error) {
+	backend := cfg.Backend
+
+	// Init the db connections for sql backends.
+	err := cfg.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	// Settings to open a particular db backend.
+	var args []interface{}
+
+	switch backend {
+	case kvdb.PostgresBackendName:
+		args = []interface{}{
+			context.Background(),
+			&postgres.Config{
+				Dsn:            cfg.Postgres.Dsn,
+				Timeout:        time.Minute,
+				MaxConnections: 10,
+			},
+			prefix,
+		}
+
+		logger.Infof("Opening postgres backend at %s with prefix '%s'",
+			cfg.Postgres.Dsn, prefix)
+
+	case kvdb.SqliteBackendName:
+		// Directories where the db files are located.
+		graphDir := lncfg.CleanAndExpandPath(
+			filepath.Join(cfg.Sqlite.DataDir, "graph", network),
+		)
+		walletDir := lncfg.CleanAndExpandPath(
+			filepath.Join(
+				cfg.Sqlite.DataDir, "chain", "bitcoin", network,
+			),
+		)
+
+		// In case the data directory was set but the watchtower is
+		// still the default one, we use the data directory for the
+		// watchtower as well.
+		towerServerDir := lncfg.CleanAndExpandPath(
+			filepath.Join(
+				cfg.Sqlite.TowerDir, "watchtower", "bitcoin",
+				network,
+			),
+		)
+		if cfg.Sqlite.DataDir != defaultDataDir &&
+			cfg.Sqlite.TowerDir == defaultDataDir {
+
+			towerServerDir = lncfg.CleanAndExpandPath(
+				filepath.Join(
+					cfg.Sqlite.DataDir, "watchtower",
+					"bitcoin", network,
+				),
+			)
+		}
+
+		var dbName string
+		var path string
+		switch prefix {
+		case lncfg.NSChannelDB:
+			path = graphDir
+			dbName = lncfg.SqliteChannelDBName
+
+		case lncfg.NSWalletDB:
+			path = walletDir
```

Comment:
Does this violate existing assumptions in lnd re if all the sub DBs have their own contained database? IIRC we ended up combining the wallet and lnd DBs.

Reply:
not sure how it vaialates the assumptions because for the sqlite case the channel.sqlite is places in the `walletDir` ?

https://github.com/lightningnetwork/lnd/blob/a7c2c47397258eac69e88d232fbf3ccc503ad308/lncfg/db.go#L527